File: gospider/ai/summarizer.go
package ai

import (
	"regexp"
	"sort"
	"strings"

	"github.com/PuerkitoBio/goquery"
)

type SummaryResult struct {
	Title   string `json:"title"`
	Summary string `json:"summary"`
	Method  string `json:"method"`
}

type ContentSummarizer struct {
	MaxSentences int
}

func NewContentSummarizer(maxSentences int) *ContentSummarizer {
	if maxSentences <= 0 {
		maxSentences = 3
	}
	return &ContentSummarizer{MaxSentences: maxSentences}
}

func (s *ContentSummarizer) Summarize(html string) SummaryResult {
	doc, err := goquery.NewDocumentFromReader(strings.NewReader(html))
	if err != nil {
		return SummaryResult{Summary: strings.TrimSpace(html), Method: "fallback"}
	}
	title := strings.TrimSpace(doc.Find("title").First().Text())
	if title == "" {
		title = strings.TrimSpace(doc.Find("h1").First().Text())
	}
	description := strings.TrimSpace(firstMetaContent(doc, "meta[name='description'], meta[property='og:description']"))
	if description != "" {
		return SummaryResult{Title: title, Summary: description, Method: "meta_description"}
	}
	text := extractMainText(doc)
	sentences := splitSentences(text)
	if len(sentences) <= s.MaxSentences {
		return SummaryResult{Title: title, Summary: strings.Join(sentences, " "), Method: "extractive"}
	}
	wordFreq := calculateWordFrequency(text)
	scored := make([]scoredSentence, 0, len(sentences))
	for index, sentence := range sentences {
		scored = append(scored, scoredSentence{
			Index:    index,
			Score:    scoreSentence(sentence, wordFreq, index, len(sentences)),
			Sentence: sentence,
		})
	}
	sort.Slice(scored, func(i, j int) bool { return scored[i].Score > scored[j].Score })
	if len(scored) > s.MaxSentences {
		scored = scored[:s.MaxSentences]
	}
	sort.Slice(scored, func(i, j int) bool { return scored[i].Index < scored[j].Index })
	parts := make([]string, 0, len(scored))
	for _, item := range scored {
		parts = append(parts, item.Sentence)
	}
	return SummaryResult{Title: title, Summary: strings.Join(parts, " "), Method: "extractive"}
}

type scoredSentence struct {
	Index    int
	Score    float64
	Sentence string
}

func firstMetaContent(doc *goquery.Document, selector string) string {
	var value string
	doc.Find(selector).EachWithBreak(func(_ int, selection *goquery.Selection) bool {
		value = strings.TrimSpace(selection.AttrOr("content", ""))
		return value == ""
	})
	return value
}

func extractMainText(doc *goquery.Document) string {
	selectors := []string{"article", ".content", ".main-content", ".article-content", ".post-content"}
	for _, selector := range selectors {
		text := strings.TrimSpace(doc.Find(selector).First().Text())
		if text != "" {
			return text
		}
	}
	return strings.TrimSpace(doc.Text())
}

func splitSentences(text string) []string {
	re := regexp.MustCompile(`[。！？.!?]+`)
	raw := re.Split(text, -1)
	sentences := make([]string, 0, len(raw))
	for _, sentence := range raw {
		trimmed := strings.TrimSpace(sentence)
		if trimmed != "" {
			sentences = append(sentences, trimmed)
		}
	}
	return sentences
}

func calculateWordFrequency(text string) map[string]int {
	re := regexp.MustCompile(`\w+`)
	stopWords := map[string]struct{}{
		"the": {}, "a": {}, "an": {}, "and": {}, "or": {}, "but": {}, "in": {}, "on": {}, "at": {}, "to": {},
		"for": {}, "的": {}, "了": {}, "是": {}, "在": {}, "我": {}, "有": {}, "和": {}, "就": {}, "不": {},
	}
	freq := map[string]int{}
	for _, word := range re.FindAllString(strings.ToLower(text), -1) {
		if len(word) <= 1 {
			continue
		}
		if _, blocked := stopWords[word]; blocked {
			continue
		}
		freq[word]++
	}
	return freq
}

func scoreSentence(sentence string, freq map[string]int, index, total int) float64 {
	re := regexp.MustCompile(`\w+`)
	words := re.FindAllString(strings.ToLower(sentence), -1)
	if len(words) == 0 {
		return 0
	}
	score := 0.0
	for _, word := range words {
		score += float64(freq[word])
	}
	if index == 0 || index == total-1 {
		score += 1
	}
	lengthPenalty := 1.0
	if len(words) > 40 {
		lengthPenalty = 0.8
	}
	return score * lengthPenalty / float64(len(words))
}

File: rustspider/src/ai/summarizer.rs
use regex::Regex;
use scraper::{Html, Selector};
use std::collections::HashMap;

#[derive(Debug, Clone, PartialEq)]
pub struct SummaryResult {
    pub title: String,
    pub summary: String,
    pub method: String,
}

pub struct ContentSummarizer {
    max_sentences: usize,
}

impl Default for ContentSummarizer {
    fn default() -> Self {
        Self { max_sentences: 3 }
    }
}

impl ContentSummarizer {
    pub fn new(max_sentences: usize) -> Self {
        Self {
            max_sentences: max_sentences.max(1),
        }
    }

    pub fn summarize(&self, html: &str) -> SummaryResult {
        let document = Html::parse_document(html);
        let title = select_text(&document, "title")
            .or_else(|| select_text(&document, "h1"))
            .unwrap_or_default();
        if let Some(description) = select_meta(
            &document,
            r#"meta[name="description"], meta[property="og:description"]"#,
        ) {
            return SummaryResult {
                title,
                summary: description,
                method: "meta_description".to_string(),
            };
        }
        let text = extract_main_text(&document);
        let sentences = split_sentences(&text);
        if sentences.len() <= self.max_sentences {
            return SummaryResult {
                title,
                summary: sentences.join(" "),
                method: "extractive".to_string(),
            };
        }
        let frequencies = calculate_word_frequency(&text);
        let mut scored: Vec<(usize, f64, String)> = sentences
            .iter()
            .enumerate()
            .map(|(index, sentence)| {
                (
                    index,
                    score_sentence(sentence, &frequencies, index, sentences.len()),
                    sentence.clone(),
                )
            })
            .collect();
        scored.sort_by(|left, right| {
            right
                .1
                .partial_cmp(&left.1)
                .unwrap_or(std::cmp::Ordering::Equal)
        });
        scored.truncate(self.max_sentences);
        scored.sort_by_key(|item| item.0);
        SummaryResult {
            title,
            summary: scored
                .into_iter()
                .map(|(_, _, sentence)| sentence)
                .collect::<Vec<_>>()
                .join(" "),
            method: "extractive".to_string(),
        }
    }
}

fn select_text(document: &Html, selector: &str) -> Option<String> {
    let selector = Selector::parse(selector).ok()?;
    document
        .select(&selector)
        .next()
        .map(|node| node.text().collect::<Vec<_>>().join(" ").trim().to_string())
        .filter(|value| !value.is_empty())
}

fn select_meta(document: &Html, selector: &str) -> Option<String> {
    let selector = Selector::parse(selector).ok()?;
    document
        .select(&selector)
        .next()
        .and_then(|node| node.value().attr("content"))
        .map(str::trim)
        .map(ToOwned::to_owned)
        .filter(|value| !value.is_empty())
}

fn extract_main_text(document: &Html) -> String {
    for selector in [
        "article",
        ".content",
        ".main-content",
        ".article-content",
        ".post-content",
    ] {
        if let Some(text) = select_text(document, selector) {
            return text;
        }
    }
    document
        .root_element()
        .text()
        .collect::<Vec<_>>()
        .join(" ")
        .trim()
        .to_string()
}

fn split_sentences(text: &str) -> Vec<String> {
    let re = Regex::new(r"[。！？.!?]+").expect("sentence regex");
    re.split(text)
        .map(str::trim)
        .filter(|sentence| !sentence.is_empty())
        .map(ToOwned::to_owned)
        .collect()
}

fn calculate_word_frequency(text: &str) -> HashMap<String, usize> {
    let stop_words = [
        "the", "a", "an", "and", "or", "but", "in", "on", "at", "to", "for", "的", "了", "是",
        "在", "我", "有", "和", "就", "不",
    ];
    let stop_words = stop_words
        .into_iter()
        .collect::<std::collections::HashSet<_>>();
    let re = Regex::new(r"\w+").expect("word regex");
    let mut frequencies = HashMap::new();
    for token in re.find_iter(&text.to_lowercase()) {
        let word = token.as_str();
        if word.len() <= 1 || stop_words.contains(word) {
            continue;
        }
        *frequencies.entry(word.to_string()).or_insert(0) += 1;
    }
    frequencies
}

fn score_sentence(
    sentence: &str,
    frequencies: &HashMap<String, usize>,
    index: usize,
    total: usize,
) -> f64 {
    let re = Regex::new(r"\w+").expect("word regex");
    let words: Vec<String> = re
        .find_iter(&sentence.to_lowercase())
        .map(|match_| match_.as_str().to_string())
        .filter(|word| word.len() > 1)
        .collect();
    if words.is_empty() {
        return 0.0;
    }
    let mut score = words
        .iter()
        .map(|word| *frequencies.get(word).unwrap_or(&0) as f64)
        .sum::<f64>();
    if index == 0 || index + 1 == total {
        score += 1.0;
    }
    if words.len() > 40 {
        score *= 0.8;
    }
    score / words.len() as f64
}
